### PR TITLE
Added timestamps to map history points

### PR DESCRIPTION
**Pull request by Sha-Darim**
## Proposed change

This PR basically adds the ability to display date/time information accompanied with history dots.
Added new fields to frontend map edit as well as additional language variables part from the actual date/time display.

The map with date/time additions.
map

New fields in map edit
map_edit

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x ] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

- [x ] The code change is tested and works locally.
- [x ] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

**Comment by homeassistant** [comment]
Hi @Sha-Darim, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;frontend#8668).

Once you do that we will be able to review and accept this pull request.

Thanks!

**Comment by spacegaier** [comment]
Please keep the `yarn.lock` file and add a screenshot to show how the UI looks after your changes.

File: src/panels/lovelace/cards/hui-map-card.ts [hunk 1]
```diff
@@ -450,15 +461,29 @@ class HuiMapCard extends LitElement implements LovelaceCard {
...
-              color: this._getColor(entityId),
-              opacity,
-              interactive: false,
-            })
-          );
+          if (this._config!.show_timestamps) {
+            mapPaths.push(
+              marker
+                .bindTooltip(
+                  format(
```
Comment (spacegaier) [inline src/panels/lovelace/cards/hui-map-card.ts:476]: We have existing helpers in src/common/datetime/* that should be reused.

Comment (Sha-Darim) [reply src/panels/lovelace/cards/hui-map-card.ts:476]: The map with date/time additions.
map

New fields in map edit
map_edit

Comment (Sha-Darim) [reply src/panels/lovelace/cards/hui-map-card.ts:476]: Will try to figure out format_date_time to replace my own variant.

**Comment by thomasloven** [comment]
It's a bit cluttered.

How about making the tooltips show on mouse-over only, but make them non-optional?

Also, isn't there a function somewhere already to format timestamps according to the language setting?

File: src/panels/lovelace/cards/hui-map-card.ts (hunk 1 above)
Comment (thomasloven) [reply src/panels/lovelace/cards/hui-map-card.ts:476]: Please add the screenshots to your PR description.

**Comment by Sha-Darim** [comment]
> It's a bit cluttered.

Yes. I am aware of that. Tried to make it a bit more code-effective. Got hung up on hui-map-card lines 432-453. I'm not quite up to speed to "modern js" and didn't know how to make it better. I would prefer to create just one array containing both LatLang and date. 

> How about making the tooltips show on mouse-over only, but make them non-optional?

The issue with the tooltips is to make them work nicely on a phone.  I tried to make them non-permanent but found them harder to use on my phone.

> Edit: Also, you move fast 😮
`device_tracker.see` is quite useful for this ;-)

**Comment by Sha-Darim** [comment]
BTW. About the date formatting. I would like to keep the map edit field. Is that possible using the suggested formatter?

**Comment by spacegaier** [comment]
> 
> 
> BTW. About the date formatting. I would like to keep the map edit field. Is that possible using the suggested formatter?

I was just thinking about that: Do we really need that input field? My idea would have been to allow the user to select from the existing `TimestampRenderingFormats` we already have so that we have a consistent timestamp rendering throughout the UI.

**Comment by Sha-Darim** [comment]
I've been thinking about the dateformatting issue.

I understand why my custom formatter field is out of place - it doesn't conform to what exist today.
But.
I believe Home Assistant lacks a good way to define date/time formats. The current formats are only useful at some occasions, but not all. I think we all could benefit from an editor where users could create and edit date/time formats. Some should still be "static" thou - like the ones we have today as several functions within Hass rely on those.

With this in place, we could give users the ability to select from the set formats. This would be useful both in editors (as an select list) as well in Lovelace UI-configuration. Fields like my custom formatter wouldn't be necessary. It would be replaced by a select list.

What do you think?

**Comment by bramkragten** [comment]
Like @spacegaier said, we have `TimestampRenderingFormats` options <https://github.com/home-assistant/frontend/blob/dev/src/panels/lovelace/components/types.ts#L10-L15>.

Would that not be enough?

**Comment by Sha-Darim** [comment]
@bramkragten 
What I'm talking about exist in Drupal (Content Management System). I believe we can benefit from something similar.
https://www.ostraining.com/blog/date-and-time-formats-in-drupal-8/

Home Assistant is a very complex piece of software where the need of information is as important as the need of controlling things. I believe we need to make it as easy to convey information as we can control devices. Date formatting is just one small part of that, but it's something :-)

This discussion has really nothing to do with my format fiend - that needs to be changed :.)

**Comment by bramkragten** [comment]
I don't think we need something like that, we probably will add some more options like we are adding for number formatting right now, but not user-defined.

Let's get back to topic now, if you want to discuss further, open a discussion :-)

**Comment by Sha-Darim** [comment]
So, I removed my own dateformatter and ended up with a map looking like this:
map2

map_edit2


Will try to commit the changed code tomorrow.

**Comment by bramkragten** [comment]
Is it now also only showing on mouse hover?

**Comment by Sha-Darim** [comment]
Yesterday, I tried to remove the `permanent: true` setting of the tooltip - but wasn't able to show the timestamps at all without it. Perhaps another approach to this is needed.
The map setting to show the timestamps is optional which would give the user the ability to select whether to display the timestamps or not. But then, this was the goal when I started working on this change - to be able to follow the tracks of a device easily. With the mouse over function in place, it wouldn't be possible to easily take a glance on the map to follow the movements of a device. It will only be "extra info" for a history dot and in that case, the option to switch this on or off would probably not be necessary either.

To enhance this, you would need to be able to switch all the timestamps on or off while viewing the map - and that would suggest layer switching. 

I believe one could add lots of additional data to the map as it is pretty simple today. Different layers could be added to be able to display different information. I took a glance at the base code for the map the other day and realized that it would demand quite a lot of work - work and discussion for another time and place 😉

**Comment by bramkragten** [comment]
But this is not very clear ;-)

image

File: src/panels/lovelace/cards/hui-map-card.ts [hunk 2]
```diff
@@ -441,6 +442,16 @@ class HuiMapCard extends LitElement implements LovelaceCard {
...
 
+        // filter location data to get correct timestamps
+        const pathDates = entityStates.reduce((accumulator: Date[], state) => {
+          const latitude = state.attributes.latitude;
+          const longitude = state.attributes.longitude;
+          if (latitude && longitude) {
+            accumulator.push(state.last_updated);
+          }
+          return accumulator;
+        }, []) as Date[];
```
Comment (bramkragten) [inline src/panels/lovelace/cards/hui-map-card.ts:453]: Just add this too the one above, no need to do it twice

File: src/panels/lovelace/cards/hui-map-card.ts [hunk 3]
```diff
@@ -450,15 +461,29 @@ class HuiMapCard extends LitElement implements LovelaceCard {
           const opacityStep = 0.8 / (path.length - 2);
           const opacity = 0.2 + markerIndex * opacityStep;
 
+          let marker = Leaflet.circleMarker(path[markerIndex], {
+            radius: 3,
+            color: this._getColor(entityId),
+            opacity,
+            interactive: false,
```
Comment (bramkragten) [inline src/panels/lovelace/cards/hui-map-card.ts:468]: If you want it to show on hover, you should set `interactive` to true.

Comment (bramkragten) [inline src/panels/lovelace/cards/hui-map-card.ts:468]:
```suggestion
            interactive: this._config!.show_timestamps,
```

File: src/panels/lovelace/cards/hui-map-card.ts [hunk 4]
```diff
@@ -450,15 +461,29 @@ class HuiMapCard extends LitElement implements LovelaceCard {
...
-          );
+          if (this._config!.show_timestamps) {
+            mapPaths.push(
+              marker
+                .bindTooltip(
+                  format(
+                    new Date(pathDates[markerIndex]),
+                    this._config!.timestamp_format
+                  ),
+                  { permanent: true }
```
Comment (bramkragten) [inline src/panels/lovelace/cards/hui-map-card.ts:480]:
```suggestion
```

**Comment by Sha-Darim** [comment]
> But this is not very clear ;-)
> 
> image

Yeah, I know :-P. That is pretty bad. My initial idéa worked only as long as the historydots were spread out., it get pretty tangled up  otherwise... I started thinking about implementing marker-stacking, but that is an idéa for another time = ). Continuing to implement the mouse-over variant instead.

**Comment by Sha-Darim** [comment]
Now, everything should be in order.

This was my first attempt to contribute with something. I guess the committed code is correct now but the way I did this was a messy. It shows I've been away from coding way too long :-D

File: src/panels/lovelace/cards/hui-map-card.ts [hunk 5]
```diff
@@ -431,26 +431,16 @@ class HuiMapCard extends LitElement implements LovelaceCard {
 
         // filter location data from states and remove all invalid locations
         const path = entityStates.reduce(
-          (accumulator: LatLngTuple[], state) => {
+          (accumulator: [LatLngTuple, Date][], state) => {
```
Comment (bramkragten) [inline src/panels/lovelace/cards/hui-map-card.ts:434]:
```suggestion
          (accumulator: {latLng: LatLngTuple, date: Date}[], state) => {
```

**Comment by bramkragten** [comment]
I still miss `yarn.lock`

File: src/panels/lovelace/cards/hui-map-card.ts (hunk 5 above)
Comment (Sha-Darim) [reply src/panels/lovelace/cards/hui-map-card.ts:434]: That looks reasonable :-)

**Comment by Sha-Darim** [comment]
> I still miss `yarn.lock`

Yes. How do you suggest I get it back? Just recommit the original file to nullify the effect?

**Comment by bramkragten** [comment]
> > I still miss `yarn.lock`
> 
> Yes. How do you suggest I get it back? Just recommit the original file to nullify the effect?

That will work

**Pull request closed without merging**